text/0000-crt-static.md

```diff
@@ -0,0 +1,371 @@
+  target-feature=-crt-static`. This extends `-C target-feature` to mean not just
+  "CPU feature" ala LLVM, but "feature of the Rust target". Several existing
+  properties of this flag, the ability to add, with `+`, _or remove_, with `-`,
+  the feature, as well as the automatic lowering to `cfg` values, are crucial to
+  later aspects of the design. This target feature will be added to targets via
+  a small extension to the compiler's target specification.
+- Lowering `cfg` values to Cargo build script environment variables. This will
+  enable build scripts to understand all enabled features of a target (like
+  `crt-static` above) to, for example, compile C code correctly on MSVC.
+- Lazy link attributes. This feature is only required by std's own copy of the
+  libc crate, and only because std is distributed in binary form and it may yet
+  be a long time before Cargo itself can rebuild std.
+
+### Specifying dynamic/static C runtime linkage
+
+A new `target-feature` flag will now be supported by the compiler for relevant
+targets: `crt-static`. This can be enabled and disabled in the compiler via:
+
+```
+rustc -C target-feature=+crt-static ...
+rustc -C target-feature=-crt-static ...
+```
+
+Currently all `target-feature` flags are passed through straight to LLVM, but
+this proposes extending the meaning of `target-feature` to Rust-target-specific
+features as well. Target specifications will be able to indicate what custom
+target-features can be defined, and most existing targets will define a new
+`crt-static` feature which is turned off by default (except for musl).
+
+The default of `crt-static` will be different depending on the target. For
+example `x86_64-unknown-linux-musl` will have it on by default, whereas
+`arm-unknown-linux-musleabi` will have it turned off by default.
+
+### Lowering `cfg` values to Cargo build script environment variables
+
+Cargo will begin to forward `cfg` values from the compiler into build
+scripts. Currently the compiler supports `--print cfg` as a flag to print out
+internal cfg directives, which Cargo uses to implement platform-specific
+dependencies.
+
+When Cargo runs a build script it already sets a [number of environment
+variables][cargo-build-env], and it will now set a family of `CARGO_CFG_*`
+environment variables as well. For each key printed out from `rustc --print
+cfg`, Cargo will set an environment variable for the build script to learn
+about.
+
+[cargo-build-env]: http://doc.crates.io/environment-variables.html#environment-variables-cargo-sets-for-build-scripts
+
+For example, locally `rustc --print cfg` prints:
+
+```
+target_os="linux"
+target_family="unix"
+target_arch="x86_64"
+target_endian="little"
+target_pointer_width="64"
+target_env="gnu"
+unix
+debug_assertions
+```
+
+And with this Cargo would set the following environment variables for build
+script invocations for this target.
+
+```
+export CARGO_CFG_TARGET_OS=linux
```

Comment:
Will these also be set when one calls e.g. `cargo rustc -- -C target-feature=+avx`? That would make these variables available for the target crate but not for its dependencies which ... might be surprising. But I guess the same already happens with `cargo rustc` and `cfg(target_feature)`?